File: Scripts/ex8_1_1.m
% exercise 8.1.1

% Load data
cdir = fileparts(mfilename('fullpath')); 
load(fullfile(cdir,'../Data/wine2'))

%% Crossvalidation
% Create crossvalidation partition for evaluation
% using stratification and 50 pct. split between training and test 
CV = cvpartition(classNames(y+1), 'Holdout', .5);

% Extract the training and test set
X_train = X(CV.training, :);
y_train = y(CV.training);
X_test = X(CV.test, :);
y_test = y(CV.test);

%% Fit model
% Fit logistic regression model to training data to predict the type of wine
w_est = glmfit(X_train, y_train, 'binomial');

% Evaluate the logistic regression on the test data
p = glmval(w_est, X_test, 'logit');

%% Plot receiver operating characteristic (ROC) curve
mfig('ROC'); clf;
rocplot(p, y_test);

% We'll plot a confusion matrix as well
mfig('Confusion matrix'); clf;
y_test_est = p>.5;
confmatplot(classNames(y_test+1), classNames(y_test_est+1));

File: Python/Scripts/ex8_1_1.py
# exercise 8.1.1

from pylab import *
from scipy.io import loadmat
from sklearn import cross_validation
from sklearn.linear_model import LogisticRegression
from toolbox_02450 import rocplot, confmatplot

# Load Matlab data file and extract variables of interest
mat_data = loadmat('../Data/wine2.mat')
X = np.matrix(mat_data['X'])
y = np.matrix(mat_data['y'], dtype=int)
attributeNames = [name[0] for name in mat_data['attributeNames'][0]]
classNames = [name[0][0] for name in mat_data['classNames']]
N, M = X.shape
C = len(classNames)

# K-fold crossvalidation
K = 2
CV = cross_validation.StratifiedKFold(y.A.ravel().tolist(),K)

k=0
for train_index, test_index in CV:

    # extract training and test set for current CV fold
    X_train, y_train = X[train_index,:], y[train_index,:]
    X_test, y_test = X[test_index,:], y[test_index,:]

    logit_classifier = LogisticRegression()
    logit_classifier.fit(X_train, y_train.A.ravel())

    y_test_est = np.mat(logit_classifier.predict(X_test)).T
    p = np.mat(logit_classifier.predict_proba(X_test)[:,1]).T

    figure(k)
    rocplot(p, y_test)

    figure(k+1)
    confmatplot(y_test,y_test_est)

    k+=2

show()
